File: src/main/java/team404/PlayerRevivalFileManager.java
package team404;

import org.apache.commons.lang3.tuple.Pair;
import org.bukkit.Material;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static team404.utils.FileUtils.appendToFile;
import static team404.utils.FileUtils.createFile;
import static team404.utils.FileUtils.readAllLines;
import static team404.utils.FileUtils.writeToFile;

public class PlayerRevivalFileManager {

    private static PlayerRevivalFileManager instance;
    private static final String PLAYERS_TO_REVIVE_FILE_NAME = "players_to_revive.txt";
    private static final String RESPAWNABLE_PLAYERS_FILE_NAME = "respawnable_players.txt";

    private final File playersToReviveFile;
    private final File respawnablePlayersFile;

    private PlayerRevivalFileManager(Plugin plugin) {
        playersToReviveFile = createFile(plugin.getDataFolder(), PLAYERS_TO_REVIVE_FILE_NAME);
        respawnablePlayersFile = createFile(plugin.getDataFolder(), RESPAWNABLE_PLAYERS_FILE_NAME);
    }

    public static PlayerRevivalFileManager getInstance(Plugin plugin) {
        if (instance == null) {
            instance = new PlayerRevivalFileManager(plugin);
        }
        return instance;
    }

    public Map<String, List<Pair<Integer, Material>>> readRespawnablePlayers() {
        return readAllLines(respawnablePlayersFile).stream()
                .map(this::processRespawnablePlayer)
                .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
    }

    public List<String> readPlayersToSpawn() {
        return readAllLines(playersToReviveFile);
    }

    private Pair<String, List<Pair<Integer, Material>>> processRespawnablePlayer(String line) {
        var keys = line.split(",");
        var playerName = keys[0];
        var materialList = Arrays.stream(keys)
                .skip(1)
                .map(a -> {
                    var b = a.split(":");
                    return Pair.of(Integer.valueOf(b[1]), Material.getMaterial(b[0]));
                })
                .collect(Collectors.toList());
        return Pair.of(playerName, materialList);
    }

    public Collection<String> respawnablePlayersToContent(Map<String, List<Pair<Integer, Material>>> map) {
        return map.entrySet().stream()
                .map(entry -> respawnablePlayersToContent(entry.getKey(), entry.getValue()))
                .toList();
    }

    public String respawnablePlayersToContent(String playerName, List<Pair<Integer, Material>> materials) {
        return playerName + "," + materials.stream()
                .map(pair -> pair.getRight() + ":" + pair.getLeft())
                .collect(Collectors.joining(",")) + "\n";
    }

    public void syncRespawnablePlayers(Map<String, List<Pair<Integer, Material>>> map) {
        writeToFile(respawnablePlayersFile, respawnablePlayersToContent(map));
    }

    public void syncPlayersToRespawn(Collection<String> playersToRevive) {
        writeToFile(playersToReviveFile, playersToRevive);
    }

    public void addRespawnablePlayer(String playerName, List<Pair<Integer, Material>> materials) {
        appendToFile(respawnablePlayersFile, respawnablePlayersToContent(playerName, materials));
    }

    public void addPlayerToRespawn(String playerName) {
        appendToFile(playersToReviveFile, playerName + "\n");
    }


}

File: src/main/kotlin/team404/service/PlayerRevivalFileManager.kt
package team404.service

import org.apache.commons.lang3.tuple.Pair
import org.bukkit.Material
import org.bukkit.plugin.Plugin
import team404.util.FileUtils.appendToFile
import team404.util.FileUtils.createFile
import team404.util.FileUtils.readAllLines
import team404.util.FileUtils.writeToFile
import java.io.File

class PlayerRevivalFileManager private constructor(plugin: Plugin) {

    private val playersToReviveFile: File = createFile(plugin.dataFolder, PLAYERS_TO_REVIVE_FILE_NAME)
    private val respawnablePlayersFile: File = createFile(plugin.dataFolder, RESPAWNABLE_PLAYERS_FILE_NAME)

    fun readRespawnablePlayers(): Map<String, Map<Material, Int>> {
        return readAllLines(respawnablePlayersFile)
            .filter { it.isNotEmpty() }
            .map { processRespawnablePlayer(it) }
            .associate { it.left to it.right }
    }

    fun readPlayersToSpawn(): List<String> {
        return readAllLines(playersToReviveFile)
    }

    private fun processRespawnablePlayer(line: String): Pair<String, Map<Material, Int>> {
        val keys = line.split(",")
        val playerName = keys[0]
        val materialMap = keys.drop(1)
            .map { it.split(":") }
            .associate { (material, count) -> Material.getMaterial(material) to count.toInt() }
            .filterKeys { it != null }
            .mapKeys { it.key!! }
            .toMutableMap()
        return Pair.of(playerName, materialMap)
    }

    private fun respawnablePlayersToContent(map: Map<String, Map<Material, Int>>): Collection<String> {
        return map.entries.map { respawnablePlayersToContent(it.key, it.value) }
    }

    private fun respawnablePlayersToContent(playerName: String, materials: Map<Material, Int>): String {
        return "$playerName," + materials.entries.joinToString(",") { "${it.key}:${it.value}" } + "\n"
    }

    fun syncRespawnablePlayers(map: Map<String, Map<Material, Int>>) {
        writeToFile(respawnablePlayersFile, respawnablePlayersToContent(map))
    }

    fun syncPlayersToRespawn(playersToRevive: Collection<String>) {
        writeToFile(playersToReviveFile, playersToRevive)
    }

    fun addRespawnablePlayer(playerName: String, materials: Map<Material, Int>) {
        appendToFile(respawnablePlayersFile, respawnablePlayersToContent(playerName, materials))
    }

    fun addPlayerToRespawn(playerName: String) {
        appendToFile(playersToReviveFile, "$playerName\n")
    }

    companion object {
        private const val PLAYERS_TO_REVIVE_FILE_NAME = "players_to_revive.txt"
        private const val RESPAWNABLE_PLAYERS_FILE_NAME = "respawnable_players.txt"

        @Volatile
        private var instance: PlayerRevivalFileManager? = null

        fun getInstance(plugin: Plugin): PlayerRevivalFileManager =
            instance ?: synchronized(this) {
                instance ?: PlayerRevivalFileManager(plugin).also { instance = it }
            }
    }
}
